[18.0][MIG] stock_picking_return_lot

StefanRijnhart opened the pull request:
Full refactoring to adapt to Odoo's refactored stock.picking.return wizard which has separate options for returning all quantities or manually entered quantities.` This module now logically integrates with stock_picking_return_restricted_qty because the restricted quantities have to be split up per lot so that the modules would be incompatible and have the duplicate functionality of setting the default quantity otherwise.

Commit:
[ADD] stock_picking_return_lot: create

renda-dev committed:
[MIG] stock_picking_return_lot: Migration to 14.0

sbejaoui committed:
[16.0][MIG] stock_picking_return_lot

sbejaoui committed:
[IMP] stock_picking_return_lot: Add a lot field to the return wizard

- Add a lot field to the return wizard and create a return line for each lot/product combination.
- Use Stock Restrict Lot to track the accepted lot on the return picking.

mymage committed:
Added translation using Weblate (Italian)

sbejaoui committed:
[FIX] stock_picking_return_lot: group return line by move and lot

In case a picking out contains multiple moves for the same product/lot,
it's essential to split the return lines by lot move to ensure the correct linking between
the return move and the delivery move.

Commit:
[IMP] stock_picking_return_lot: pre-commit auto fixes

Commit:
[MIG] stock_picking_return_lot: Migration to 17.0

StefanRijnhart commented:
/ocabot migration stock_picking_return_lot

sbejaoui reviewed (approved):
code review + functional test

marielejeune commented on stock_picking_return_lot/tests/test_stock_picking_return_lot.py:
```diff
@@ -55,11 +64,9 @@ def setUpClass(cls):
-            .create({})
-        )
+        wizard = cls.env["stock.return.picking"].create({"picking_id": picking.id})
+        # wizard.set_quantities()
```
Could you remove this commented line?

StefanRijnhart replied:
Thanks, done.

marielejeune commented on stock_picking_return_lot/wizards/stock_return_picking_line.py:
```diff
@@ -7,14 +7,49 @@
+        Overwrite of this method from stock_picking_return_restricted_quantity.
+        """
+        if self:
+            self.ensure_one()
```
Isn't this strange to have a `self.ensure_one()` on an `api.model` method?

StefanRijnhart replied:
I can see why you'd think that! But it is simply the case that `stock_picking_return_restricted_qty` calls this method on both empty recordsets and single records. For our purposes (filtering the quantities by lot) it works either way but not if someone decides to call it on two records or more so I'm just making things explicit here.

The decorator might just as well be removed from a technical point of view, because this method is only going to be called from code and not from UI, cron or RPC so the module would still be working fine without it.

marielejeune replied:
Thanks for the explanation!

StefanRijnhart committed:
[MIG] stock_picking_return_lot: Migration to 18.0

Full refactoring to adapt to Odoo's refactored stock.picking.return wizard
which has separate options for returning all quantities or manually entered
quantities. This module now logically integrates with
stock_picking_return_restricted_qty because the restricted quantities have to
be split up per lot so that the modules would be incompatible and have the
duplicate functionality of setting the default quantity otherwise.

marielejeune reviewed (approved):
LGTM

sbejaoui commented:
/ocabot merge nobump
